### PR TITLE
[PGNCCL] Use long/short wait for different non-blocking calls

### PR DESCRIPTION
**Pull request by kwen2501**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #142291

In nonblocking mode, we always check if the NCCL communicator is ready between issuing commands to it.
Today this is done by the `waitReady()` function. 
Unfortunately, the `waitReady()` function is burned with `C10D_NCCL_CHECK_TIMEOUT_SLEEP` which would sleep for an interval between two consecutive checks.
While this is nice when waiting for comm init or finalize, it degrades performance of collective calls (which would almost certainly return success immediately.)

This PR adds a `bool longInterval` argument to `waitReady` and let call site determine whether long wait is likely; if not, `waitReady` would use `sched_yield()` to more eagerly check for readiness. 

Thanks @eqy for reporting an issue that small collectives has perf impact in nonblocking mode.

cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o

### CODE REVIEW

**Review by eqy — APPROVED** [review]
approving but the failures look real

**Comment by kwen2501** [comment]
Thanks @eqy , yeah, just fixed them

File: torch/csrc/distributed/c10d/NCCLUtils.cpp [hunk 1]
```diff
@@ -25,7 +25,9 @@ ncclComm_t NCCLComm::getNcclComm() {
   }
   // In non-blocking mode, ensure comm is ready.
   if (nonBlocking_) {
-    waitReady();
+    // Wait with long interval if communicator is being initialized.
+    bool longInterval = !initialized_;
+    waitReady(longInterval);
```
Comment (fduwjj) [inline torch/csrc/distributed/c10d/NCCLUtils.cpp:30]: can you maybe directly pass in `(!initialized_)`?

**Review by fduwjj — APPROVED** [review]

**Pull request closed without merging**